speedup cmake checks

arogge opened the pull request:
This PR tries to refactor a lot of CMake checks to improve configuration speed.

First and foremost it consolidates "does a function exist" checks, as running the compiler over and over takes a lot of time.
It also removes checks that are useless nowadays and skips platform-dependant checks on platforms where they won't succeed.
We also adds a shortcut for determining supported compiler-flags by shipping a list of presets.
As modern compilers can check if an include exists using `__has_include()`, most of the `HAVE_XXX_H` as not needed anymore and could be removed.

Lastly, a lot of other `config.h` macros have been merged while others have been removed completely.

On my machine the initial `cmake` run is down from ~15 seconds to ~6 seconds.

### Thank you for contributing to the Bareos Project!

#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

##### TODO
- [x] Fix xattr detection on MacOS
- [x] Fix extattr detection on FreeBSD
- [x] Fix setting `HAVE_XATTR`
- [x] Fix #2488 
- [x] use file-prefix-map instead of debug/macro-prefix-map
- [x] Add fcanon-prefix-map (if possible) to fix our debug paths.

### Checklist for the _reviewer_ of the PR (will be processed by the Bareos team)
Make sure you check/merge the PR using `devtools/pr-tool` to have some simple automated checks run and a proper changelog record added.

##### General
- [x] Is the PR title usable as CHANGELOG entry?
- [x] Purpose of the PR is understood
- [x] Commit descriptions are understandable and well formatted
~~Required backport PRs have been created~~
~~Correct milestone is set~~

##### Source code quality
- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR

arogge committed:
cmake: refactor BareosCheckFunctions

* reorder checks by topic
* merge checks that succeed on all platforms into a single check

arogge committed:
cmake: remove unused PROPLIST defines

arogge committed:
cmake: refactor linux xattr checks

this now uses a single try_compile to check for all required functions
and headers at once.

arogge committed:
cmake: assume POSIX headers are present on non-MSVC

arogge committed:
cmake: remove pointless AFS include-checks

when HAVE_AFS_ACL is enabled and one of the two headers we check for is
missing, the build will fail. As it will also fail if we just don't
check for the headers, the check itself is pointless and can be removed.

arogge committed:
cmake: move FreeBSD and AIX extended attribute checks

moves checks for FreeBSD and for AIX into merged compile tests.

arogge committed:
cmake: unify naming of check_*_compiler_flag results

arogge committed:
cmake: short-cut known compiler-flags

add a new cmake module that knows about the compiler flags that we use
and which compiler supports them. Thus, instead of running the compiler
over and over during configuration-phase we pre-seed the flag variables.

arogge committed:
glfs-fd: remove compatiblility with gfapi < 3.5

arogge committed:
cmake: remove BareosCheckSymbols

removes the stub-checks from CMake and add them to the `#if`s in the
code.

arogge committed:
aix: merge EA defines

arogge committed:
linux/macos: merge xattr defines

Linux and MacOS both use getxattr()/setxattr() but there are subtle
differences. Thus we need to check for both individually.

arogge committed:
freebsd: merge extattr defines

arogge committed:
remove HAVE_SYS_TIME_H

this is true for all our platforms, so no conditional needed.
Also TIME_WITH_SYS_TIME is never set, so we can remove that, too.

arogge committed:
remove HAVE_INET_NTOP

arogge committed:
remove HAVE_GETADDRINFO

arogge committed:
remove HAVE_GLOB

arogge committed:
remove HAVE_LCHOWN

arogge committed:
cmake: limit os-specific compile-tests to each os

as it is pointless to check for platform-specific features on every
platform, we just run the tests when building the specific platform.

sebsura commented on core/src/compile_tests/linux_xattr.c:
```diff
@@ -0,0 +1,42 @@
+  (void)argv;
+
+  // Linux extended attribute functions
+  getxattr(ptr, ptr2, ptr3, 0);
```
but i guess this way its more resilient to small changes (i.e. char* instead of const char*).

arogge replied:
* using `static_assert()` requires the C++ compiler
* functions to be checked can be (arbitrarily complex) macros
* this batches `check_function_exists()`, so I mimicked the code that CMake emits for that

TL;DR I wanted these checks to be as identical to the original one as possible

arogge commented on cmake/BareosKnownCompilerFlags.cmake:
```diff
@@ -0,0 +1,147 @@
+        set(supported "")
+      endif()
+
+      # fixup poorly chosen name: while the CMake LANG is "C", our
+      # cache-variables start with "CC"
+      if(lang STREQUAL "C")
```
we probably should, as I wrote above that's backwards compatibility to my own stupidity :)

arogge commented on core/CMakeLists.txt:
```diff
@@ -109,43 +109,28 @@ endif()
 
-check_cxx_compiler_flag(-Winvalid-offsetof compiler_has_invalid_offsetof)
-if(compiler_has_invalid_offsetof)
+check_cxx_compiler_flag(-Winvalid-offsetof CXX_SUPPORTS_Winvalid_offsetof)
+if(CXX_SUPPORTS_Winvalid_offsetof)
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-invalid-offsetof")
 endif()
```
Not really. The checks sees if some argument to `-W` exists (i.e. if the compiler knows that warning). If the compiler knows it that implies we can inverse it with `-Wno`, make it an error with `-Werror=` or make it no error with `-Wno-error=`.

arogge commented on core/CMakeLists.txt:
```diff
@@ -109,43 +109,28 @@ endif()
   # make format-security issues a compiler-error
-  check_cxx_compiler_flag(-Wformat compiler_format_security)
-  if(${compiler_format_security})
+  check_cxx_compiler_flag(-Wformat CXX_SUPPORTS_Wformat)
+  if(CXX_SUPPORTS_Wformat)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wformat")
   endif()
```
I guess we need something like
`bareos_check_compiler_flags(<list-of-flags> C_ONLY <list-of-flags> CXX_ONLY <list-of-flags>)`
which iterates over the lists and runs `check_c_complier_flag()`/`check_cxx_compiler_flag()` with automatically generated cache variable names first.

And then we could have `bareos_enable_global_warnings()` analog to `bareos_disable_warnings()` so that you just do `bareos_enable_global_warnings(-Wformat -Wformat-security)`

arogge replied:
I wonder if we should use `add_compile_options()` instead of mutating `CMAKE_<LANG>_FLAGS`.

arogge commented on core/CMakeLists.txt:
```diff
@@ -223,9 +208,6 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "FreeBSD")
   include_directories(SYSTEM /usr/local/include)
   link_directories(/usr/local/lib)
   link_libraries(intl)
-  check_cxx_compiler_flag(
-    -Wunused-but-set-variable compiler_will_warn_of_unused_but_set_variable
-  )
```
That is already enabled with `-Wall` or `-Wunused`. We just need to know if it exists so we can disable it for some targets.

arogge committed:
cmake: refactor compiler flag checks

* add `bareos_check_compiler_flags()`
  this will check for support of a list of flags in the compilers and
  set cache variables with pre-determined names for these
* add `bareos_add_compile_flags()`
  this will add (global) compile flags, but filters out all unsupported
  flags.
* rename CC_SUPPORTS_* -> C_SUPPORTS_*
* reorganize warning settings in core/CMakeLists.txt

arogge committed:
cmake: BareosDisableWarnings -> BareosCompilerFlags

arogge committed:
lib: refactor drop() privileges

* stop compiling an empty stub for windows
* skip calling drop() on windows
* remove now unused HAVE_GRP_H/HAVE_PWD_H

arogge committed:
cmake: HAVE_POLL_H/HAVE_SYS_POLL_H -> HAVE_POLL

Every platform with HAVE_SYS_POLL_H also sets HAVE_POLL_H. So that
fallback is not needed anymore. Also, there is HAVE_POLL that tells us
if poll() is available, so we re-use that flag.

arogge committed:
cmake: remove unused HAVE_SYS_STATVFS_H

arogge committed:
cmake: remove HAVE_FCHMOD/HAVE_FCHMOD/HAVE_FUTIMENS

arogge committed:
cmake: remove HAVE_UTIMES/HAVE_LUTIMES

utimes() is used as a fallback for lutimes(), but that fallback is never
required. Win32 uses neither, so we can just use HAVE_WIN32 here.

arogge committed:
cmake: remove HAVE_FCHOWNAT/HAVE_FUTIMES/HAVE_FUTIMESAT

these were only used to make the Solaris build fail. As it would also
fail if these are just not present, we can safely remove them.

arogge committed:
cmake: remove HAVE_SETREUID

priv.cc is not compiled for Win32 anymore and all other platforms have
this (as it is part of the POSIX system interfaces).

arogge committed:
cmake: remove HAVE_FDATASYNC

the fdatasync() macro we define if we don't have the function is never
used, so it can be removed.

arogge committed:
cmake: remove HAVE_FSEEKO

arogge committed:
cmake: remove HAVE_GETHOSTBYNAME2

this is never used.

arogge committed:
cmake: remove HAVE_LOCALTIME_R

we only need to distinguish Win32 and POSIX, where Win32 uses
localtime_s() and POSIX uses localtime_r().

arogge committed:
cmake: remove HAVE_NANOSLEEP

arogge committed:
cmake: remove unused HAVE_OPENAT

arogge committed:
cmake: remove unused HAVE_UNLINKAT

arogge committed:
cmake: fix ENABLE_CAPABILITY default

arogge committed:
systemtests: dbHost length check WARNING to INFO

arogge committed:
cmake: remove HAVE_COMPRESS_BOUND

arogge committed:
cmake: remove HAVE_ENGINE_LOAD_PK11

arogge committed:
cmake: remove HAVE_LIBUTIL

only HAVE_LIBUTIL_H is ever checked, so HAVE_LIBUTIL can be removed.

arogge committed:
cmake: remove HAVE_LONG_DOUBLE

arogge committed:
cmake: remove PostgreSQL conditionals

arogge committed:
cmake: use __has_include(<pthread_np.h>)

arogge committed:
cmake: replace HAVE_XXX_H with __has_include(<XXX>)

use the compiler to check if a header exists during the build instead of
determining this with CMake. This is not feasible for all cases, as some
decisions in CMake are based on that information.

arogge committed:
cmake: HAVE_REGEX_H -> __has_include(<regex.h>)

arogge committed:
cmake: refactor HAVE_LOWLEVEL_SCSI_INTERFACE

* centralize header checks
* only check relevant headers (per system type)

arogge committed:
cmake: add more known compiler flags

sebsura commented on core/src/findlib/create_file.cc:
```diff
@@ -355,7 +355,7 @@ int CreateFile(JobControlRecord* jcr,
                         attr->ofname, attr->olname, be.bstrerror());
                   return CF_ERROR;
-#  ifdef HAVE_CHFLAGS
+#  if defined(HAVE_CHFLAGS) && !defined(__stub_chflags)
```
See my comment to lchmod

sebsura commented on core/src/findlib/attribs.cc:
```diff
@@ -237,7 +227,7 @@ static inline bool RestoreFileAttributes(JobControlRecord* jcr,
 
   // Reset file times.
-#if defined(HAVE_FUTIMES)
+#if !defined(HAVE_WIN32)
```
We have this construct 3+ times in our code.  It is probably a good idea to just create a function for this that has this check once. Not in this pr though; this should be done in the spring cleaning.

arogge replied:
I can't find the other two. If you can remember them, feel free to tell me where to find them.

sebsura replied:
I should have been more specific with "this construct", as I have no idea what i was talking about.  Ill try to understand what i wanted to say.

sebsura replied:
Ill mark this as resolved, because I am not sure what the problem was.

sebsura commented on core/src/findlib/xattr.cc:
```diff
@@ -628,17 +611,12 @@ static BxattrExitCode (*os_parse_xattr_streams)(JobControlRecord* jcr,
-        || (!defined(HAVE_GETXATTR) && !defined(HAVE_LGETXATTR)) \
-        || (!defined(HAVE_SETXATTR) && !defined(HAVE_LSETXATTR))
+#    if (defined(HAVE_DARWIN_OS) && !defined(HAVE_DARWIN_XATTR)) \
+        || (defined(HAVE_LINUX_OS) && !defined(HAVE_LINUX_XATTR))
```
I would also be fine to just remove this entirely and have this at the top of the file:
```
#if !defined(HAVE_XATTR)
#  error "Missing full support for the XATTR functions."
#endif
```

sebsura commented on cmake/BareosCompilerFlags.cmake:
```diff
@@ -0,0 +1,136 @@
+  set(CMAKE_C_FLAGS
+      "${CMAKE_C_FLAGS} ${c_flags_str}"
+      PARENT_SCOPE
+  )
```
Wouldnt it be better to use

```
add_compile_options($<$<COMPILE_LANGUAGE:CXX>:${cxx_flags_str}>
                    $<$<COMPILE_LANGUAGE:C>:${c_flags_str}>)
```

if possible we should stop setting/accessing `CMAKE_` variables directly.

arogge replied:
You're right. I at least added a TODO comment for future me.

sebsura commented on cmake/BareosKnownCompilerFlags.cmake:
```diff
@@ -0,0 +1,173 @@
+      endif()
+
+      foreach(flag IN LISTS ARG_FLAGS)
+        string(REGEX REPLACE "^-" "" raw_flag "${flag}")
+        string(REPLACE "-" "_" mangled_flag "${raw_flag}")
```
Why do we have 3 functions that all implement this ? Please extract that in a function, because otherwise we have to remember to always keep that in sync.

With that said: why are they not in sync ?
```
        string(REGEX REPLACE "^-" "" raw_flag "${flag}")
        string(REPLACE "-" "_" mangled_flag "${raw_flag}")
```

vs

```
    string(REGEX REPLACE "^-" "" plain_flag ${flag})
    string(REGEX REPLACE "[-=]" "_" plain_flag_sym ${plain_flag})
```
vs
```
    string(REGEX REPLACE "^W(no-)?(error=)?" "W" plain_flag ${tmp1})
    string(REGEX REPLACE "[-=]" "_" plain_flag_sym ${plain_flag})
```

surely these should all agree ?  Otherwise the caching becomes somewhat ineffective.

arogge replied:
The first two could be merged, but then that function would have to go into a separate file, which I didn't think was worth the effort.

The third one is special: When you want to add `-Wno-xyz`, we check if `-Wxyz` is supported (to make sure every warning-flag is only added to the cache once), but then add `-Wno-xyz`.

sebsura replied:
Could you add the last part to a comment in BareosCompilerFlags ?

sebsura commented on cmake/PrefixMap.cmake:
```diff
@@ -23,12 +23,12 @@ include(CheckCXXCompilerFlag)
 check_cxx_compiler_flag(
-  -fdebug-prefix-map=${BAREOS_PREFIX_MAP} cxx_compiler_debug_prefix_map
+  -fdebug-prefix-map=${BAREOS_PREFIX_MAP} CXX_SUPPORTS_fdebug_prefix_map
 )
```
Why does this not use the new `bareos_check_compiler_flag` mechanism ?

arogge replied:
see below

sebsura commented on cmake/PrefixMap.cmake:
```diff
@@ -23,12 +23,12 @@ include(CheckCXXCompilerFlag)
+if(C_SUPPORTS_fdebug_prefix_map AND CXX_SUPPORTS_fdebug_prefix_map)
   set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fdebug-prefix-map=${BAREOS_PREFIX_MAP}")
   set(CMAKE_CXX_FLAGS
       "${CMAKE_CXX_FLAGS} -fdebug-prefix-map=${BAREOS_PREFIX_MAP}"
```
this could be done via `add_compile_flag` as well.

sebsura commented on core/CMakeLists.txt:
```diff
@@ -275,12 +234,16 @@ if(MSVC)
   )
+  list(JOIN BAREOS_WARNING_FLAGS " " BAREOS_WARNING_FLAGS_STRING)
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${BAREOS_WARNING_FLAGS_STRING}")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${BAREOS_WARNING_FLAGS_STRING}")
```
```suggestion
  add_compile_options(${BAREOS_WARNING_FLAGS_STRING}")
```

arogge replied:
I don't think this will currently work, as the flag checker does not work with MSVC (yet). I'd rather not touch this right now.

sebsura commented on core/cmake/BareosCheckFunctions.cmake:
```diff
@@ -18,82 +18,80 @@
+    message(
+      SEND_ERROR
+        "core/src/compile_tests/core_system_interfaces.c failed to build"
+    )
```
```suggestion
  try_compile(
    HAVE_CORE_SYSTEM_INTERFACES ${CMAKE_BINARY_DIR}/compile_tests
    ${PROJECT_SOURCE_DIR}/src/compile_tests/core_system_interfaces.c
    OUTPUT_VARIABLE build_output
  )
  if(HAVE_CORE_SYSTEM_INTERFACES)
    set(HAVE_GETPAGESIZE 1)
    set(HAVE_POLL 1)
    set(HAVE_READDIR_R 1)
  else()
    message(
      SEND_ERROR
        "core/src/compile_tests/core_system_interfaces.c failed to build: ${build_output}"
    )
```

if i understand the docs right, this should help in debugging compilation issues.
```
OUTPUT_VARIABLE <var>

    Store the output from the build process in the given variable.
```

sebsura reviewed (changes requested):
Looks great so far.  I added some comments, but you also need to rebase your changes onto master.

Some things that we may want to add to this pr:
* recently we merged a pr that fixes the old declarations in libdroplet.  Maybe we can built in now without disabling `-Wold-style-definition`
* Add fcanon-prefix-map (if possible) to fix our debug paths.

arogge committed:
cmake: make use of -ffile-prefix-map

instead of setting -fdebug-prefix-map and -fmacro-prefix-map, just
set -ffile-prefix-map instead.

arogge commented on cmake/PrefixMap.cmake:
```diff
@@ -37,12 +37,12 @@ if(c_compiler_debug_prefix_map AND cxx_compiler_debug_prefix_map)
 check_cxx_compiler_flag(
-  -fmacro-prefix-map=${BAREOS_PREFIX_MAP} cxx_compiler_macro_prefix_map
+  -fmacro-prefix-map=${BAREOS_PREFIX_MAP} CXX_SUPPORTS_fmacro_prefix_map
 )
```
I changed this to use `-ffile-prefix-map` now. Do we also need `-fcanon-prefix-map`?

arogge replied:
Also added `-fcanon-prefix-map`

arogge committed:
cmake: error when scsi-crypto enabled on Windows/MacOS

instead of just ignoring the flag, raise an error.

arogge committed:
cmake: output why core_system_interfaces.c fails

arogge committed:
findlib: drop AFS ACL support

this was not configurable and thus not built since moving to CMake with
Bareos 18.2.

arogge commented on core/src/stored/stored.h:
```diff
@@ -31,16 +31,16 @@
 const int sd_debuglevel = 300;
 
-#ifdef HAVE_MTIO_H
+#if __has_include(<mtio.h>)
```
I prefer the `#if __has_include(<...>)`. As it is done directly by the compiler, you can be sure there was no strange reason the definition was not there.

arogge committed:
remove options --user/--group from Win32 build

these have no effect anyways

arogge commented on core/src/lib/priv.cc:
```diff
@@ -22,16 +22,13 @@
-    && defined(HAVE_SYS_CAPABILITY_H) && defined(HAVE_PRCTL) \
-    && defined(HAVE_SETREUID)
+#if defined(ENABLE_CAPABILITY) && __has_include(<sys/prctl.h>) \
+    && __has_include(<sys/capability.h>) && defined(HAVE_PRCTL)
```
That's one of the major drawbacks of `check_function_exists()`: it only acts on the linker and tells you if that function is linkable. It doesn't check if there is a symbol for it.

I think the proper way would be to just check `#if defined(ENABLE_CAPABILITY)` and have the build fail if required headers are missing. However, we can do this in a more or less nice way. We could either `#error` in case one of the headers is missing or just let it fail. I'd probably go for the latter.

arogge committed:
privs: assume prctl() and headers are available

assume prctl() is available when ENABLE_CAPABILITY is set.

arogge commented on core/src/lib/priv.cc:
```diff
@@ -130,17 +126,14 @@ void drop(char* uname, char* gname, bool keep_readall_caps)
     cap_free(caps);
-#  else
+#else
     Emsg0(M_ERROR_TERM, 0,
           T_("Keep readall caps not implemented this OS or missing "
              "libraries.\n"));
```
That could be a bit misleading, as a user might think the feature was removed (or never existed). Not sure to be honest. The user/group on Windows was pretty clear, but this is not that obvious to me.

sebsura replied:
Makes sense.

arogge commented on core/src/findlib/attribs.cc:
```diff
@@ -57,14 +57,8 @@ static bool set_win32_attributes(JobControlRecord* jcr,
-// For old systems that don't have lchmod() use chmod()
-#ifndef HAVE_LCHMOD
+// For old systems that don't have lchmod() or where it is a stub use chmod()
+#if !defined(HAVE_LCHMOD) || defined(__stub_lchmod) || defined(__stub___lchmod)
```
So all three cases where we need it check for `HAVE_LCHMOD` and the stub macros (that have always been macros).
There is also no point in having it switchable. If `lchmod()` is a stub (like it is on Linux) it will unconditionally fail with `ESYS`.

arogge replied:
Delegating the check for `__stub_lchmod` into `config.h` is also brittle, as it would require to always including `sys/stat.h` *before* `config.h`.

arogge committed:
findlib: fix typo in AIX xattr code

arogge commented on cmake/BareosCompilerFlags.cmake:
```diff
@@ -0,0 +1,136 @@
+       AND (ARG_C_ONLY OR CXX_SUPPORTS_W${plain_warning_sym})
+    )
+      foreach(target ${ARG_TARGETS})
+        target_compile_options("${target}" PRIVATE "-Wno-${plain_warning}")
+        if(ARG_ADD_FOR_LTO)
+          target_link_options("${target}" PRIVATE "-Wno-${plain_warning}")
+        endif()
```
Usually (like >99% of the cases) you don't need the warning options during linking. I think it is (currently) unused.

arogge commented on core/src/include/config.h.in:
```diff
@@ -71,34 +68,16 @@
 #cmakedefine HAVE_AFS @HAVE_AFS@
 
 // Andrew FileSystem ACL support
 // FIXME: this is used, but never configured
```
I dropped the whole AFS ACL support. It hasn't been built for ages and is pretty easy to bring back.

arogge commented on core/cmake/BareosCheckFunctions.cmake:
```diff
@@ -18,82 +18,80 @@
-  check_function_exists(inet_ntop HAVE_INET_NTOP)
+  check_function_exists(prctl HAVE_PRCTL)
+
+  # Missing on older Linux
```
glibc older than 2.3.2 (around 2002) lacks `lchmod()`, 2.3.2 up to 2.32 (around 2020) contains the stub, glibc 2.32 onwards supports `lchmod()` by calling `fchmodat(... AT_SYMLINK_NOFOLLOW)` and Linux 6.5 (2023) finally added native support via `fchmodat2()`.
Nevertheless, I don't know if anything we're building for that is not Linux lacks an `lchmod()` call.

arogge committed:
stored: fix unreachable cmake

Fixes #2488

Remove HAVE_DARWIN_OS conditional block inside of NOT HAVE_DARWIN_OS
conditional block.

arogge committed:
cmake: add -fcanon-prefix-map if possible

add -fcanon-prefix-map when the compiler supports it and
-ffile-prefix-map was enabled, too.

arogge committed:
findlib: remove unreachable conditional code

arogge committed:
cmake: bareos_disable_warnings(): remove TARGET option

remove TARGET as we have TARGETS and a single target can be provided as
a list just as good.

arogge commented on cmake/BareosCompilerFlags.cmake:
```diff
@@ -0,0 +1,136 @@
+      bareos_check_compiler_flags("-W${plain_warning}")
+    endif()
+
+    if((ARG_CXX_ONLY OR C_SUPPORTS_W${plain_warning_sym})
+       AND (ARG_C_ONLY OR CXX_SUPPORTS_W${plain_warning_sym})
+    )
```
I guess the proper fix would be to always just have generator expressions for the supported languages (i.e. `$<$<COMPILE_LANGUAGE:C,CXX>:${flag}>`).

arogge committed:
cmake: guard try_compile() calls

guard the try_compile() calls with their result variable, so they are
not re-run on every cmake run.

arogge committed:
cmake: add guards to BareosGetDistInfo

cache results as internal variables and don't rerun the checks
everytime.

arogge committed:
cmake: remove minio certificate generator

the resulting files are not used anymore.

arogge committed:
cmake: cache more check results

* check_pymodule_available
* pam_exec

arogge committed:
cmake: disable scsi-crypto on WIN32 and MacOS

arogge committed:
cmake: set HAVE_<OS> variables early in toplevel

include BareosOSInfo early in toplevel CMakeLists.txt so the variables
are (almost) always available. Remove redundant settings of these
variables from core/CMakeLists.txt.

arogge committed:
cmake: re-add Wformat-security flags for Clang

those were lost in a conflict resolution, but needed to build on
FreeBSD.

arogge committed:
cmake: stop mutating CMAKE_{C,CXX}_FLAGS

this is considered bad practice, so it is removed where possible.
Currently the coverage analysis still seems to require it, as the order
of the commandline parameters seems to be crucial there.

arogge committed:
cmake: use COMPILE_WARNING_AS_ERROR instead of -Werror

let CMake enable -Werror and allow the user to disable it if desired
using --compile-no-warning-as-error.
While this requires CMake 3.24 or later to actually work, it won't break
any built with pre-3.24 CMake. However, warnings will not be treated as
errors in that case.

arogge commented:
I now also took the time to make it so we don't touch `CMAKE_C_FLAGS`/`CMAKE_CXX_FLAGS` (besides the Coverage stuff, that probably can't live without it currently)
I also removed `-Werror` in favour of `CMAKE_COMPILE_WARNING_AS_ERROR`.

sebsura commented on core/CMakeLists.txt:
```diff
@@ -63,34 +64,31 @@ if(MSVC)
   )
 else()
-  set(BAREOS_WARNING_FLAGS -Werror -Wall -Wextra -Wshadow)
+  set(CMAKE_COMPILE_WARNING_AS_ERROR ON) # sadly we cannot switch this on for MSVC yet
```
Please do not do this unconditionally.  This is the one thing that currently annoys me the most during development (e.g. I want to test something, but the build does not compile because now a parameter is unused ....)

```
if(NOT DEFINED CMAKE_COMPILE_WARNING_AS_ERROR)
    set(CMAKE_COMPILE_WARNING_AS_ERROR ON)
endif()
```

Then at least I can turn it off when i try stuff out.

sebsura replied:
https://github.com/gentoo/kde/blob/74c78b1a60383a136651d26f5bc8ba7899ace485/eclass/cmake.eclass#L420

```
		set(CMAKE_COMPILE_WARNING_AS_ERROR ON CACHE BOOL "")
```
apparently this works too.

arogge replied:
I tried it and what I did is the proper way to do this:
CMake supports the commandline-option [`--compile-no-warning-as-error`](https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-compile-no-warning-as-error) to ignore that flag.
It would be preferable to support both, a CACHE variable and that commandline option. 
However, when I added it as a CACHE variable, the commandline parameter did not work for me anymore and at least CMake 3.26.5 does still compile warnings as errors.

Thus, as the (otherwise pretty nice) CACHE variable solution breaks the documented CMake behaviour, I decided to do it this way.

sebsura replied:
Makes sense.  Ill add that option to my scripts.

sebsura commented on core/cmake/BareosCheckFunctions.cmake:
```diff
@@ -18,82 +18,90 @@
+      ${PROJECT_SOURCE_DIR}/src/compile_tests/core_system_interfaces.c
+      OUTPUT_VARIABLE build_output
+    )
+  endif()
```
It looks like currently every `try_compile` uses this pattern.  Should we just create a macro/function for this ?

sebsura commented on core/cmake/BareosDetermineHaveLowLevelScsiInterface.cmake:
```diff
@@ -19,77 +19,38 @@
 # determine value of HAVE_LOWLEVEL_SCSI_INTERFACE
 
-if(${scsi-crypto})
+if(scsi-crypto)
+  if(HAVE_WIN32 OR HAVE_DARWIN_OS)
+    message(FATAL_ERROR "scsi-crypto is not supported on Windows or MacOS")
+  endif()
```
Is this still relevant now that scsi-crypto is a dependant option ?

arogge replied:
you could still just blindly pass `-Dscsi-crypto=yes` to cmake, but we probably can remove that and let the build fail in that case.

sebsura commented on core/cmake/BareosGetDistInfo.cmake:
```diff
@@ -45,5 +51,14 @@ set(BAREOS_PLATFORM ${PLATFORM})
 
 # cmake_host_system_information() was added in 3.22
 if(CMAKE_VERSION VERSION_GREATER 3.22)
```
https://cmake.org/cmake/help/latest/command/cmake_host_system_information.html

 I guess the `DISTRIB_` values were meant with "added in 3.22"

sebsura replied:
The ci-builds succeeded when i set cmake_minumum to 3.22, so it probably makes sense to do that and remove this if.

arogge replied:
hmm... we require 3.25 on Windows. Only Ubuntu 22 is still on 3.22. Maybe we can upgrade that and unconditionally require 3.25 then.

sebsura commented on core/src/console/console.cc:
```diff
@@ -95,7 +95,11 @@ static int SleepCmd(FILE* input, BareosSocket* UA_sock);
+#  include <regex.h>
+#else
+#  include "lib/bregex.h"
+#endif
```
I do not get this change.  The content of `lib/bregex.h` is
```
#if __has_include(<regex.h>)
#  include <regex.h>
#elif __has_include(<pcre2posix.h>)
#  include <pcre2posix.h>
#else
#  error "no suitable regex engine found."
#endif
```

Iirc I added this header so that files need regex do not have to do this if/else dance.

sebsura replied:
Instead of the `#error` case here, we should probably just do a try-compile during the configuration.

arogge replied:
I really try not to have too many compile tests, especially not for these corner cases.
We can easily just do this:
```
#if __has_include(<regex.h>)
#  include <regex.h>
#else
#  include <pcre2posix.h>
#endif
```
And have the compiler fail. However, that's basically what we do right now - just a bit less explicit.

sebsura commented on core/src/dird/dird.cc:
```diff
@@ -48,11 +48,17 @@
+#  include <regex.h>
+#else
+#  include "lib/bregex.h"
+#endif
```
Same here

sebsura commented on core/src/lib/compression.cc:
```diff
@@ -46,11 +46,6 @@
 
 #include "fastlz/fastlzlib.h"
 
-#ifndef HAVE_COMPRESS_BOUND
-#  define compressBound(sourceLen) \
-    (sourceLen + (sourceLen >> 12) + (sourceLen >> 14) + (sourceLen >> 25) + 13)
-#endif
-
```
What about the compressBound in backup.cc ?

arogge replied:
This was dead code removed, the code in backup.cc is not dead :)

The code in `compression.cc` was never used when building with zlib (which is always the case for us). The one in backup.cc is always used, because `zlib.h` is not included and even if it was `compressBound` is not a macro.

Not sure if I want to take care of this right here and now.

sebsura replied:
No, leave it like it is.  I misunderstood this removal to mean that `compressBound` always exists anyways.

sebsura commented on systemtests/cmake/BareosSystemtestFunctions.cmake:
```diff
@@ -512,8 +512,8 @@ macro(prepare_testdir_for_daemon_run)
-      WARNING "'${dbHost}' is too long.  Creating '/tmp/${TEST_NAME}' instead."
+    message(INFO
+            "'${dbHost}' is too long.  Creating '/tmp/${TEST_NAME}' instead."
     )
```
Please describe how this was addressed.  I do not see it.

sebsura commented on core/CMakeLists.txt:
```diff
@@ -143,8 +131,7 @@ option(ENABLE_STATIC_RUNTIME_LIBS "Link C and C++ runtime libraries statically"
   if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -static-libgcc -static-libstdc++")
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -static-libgcc")
+    add_link_options(-static-libgcc -static-libstdc++)
```
Does cmake know not to set `-static-libstdc++` for C builds ?  If not then we should probably use a generator expression here.

arogge replied:
Good catch. At least `gcc` doesn't care and accepts the flag in C mode, too. Nevertheless, I added the proper generator expression now.

sebsura reviewed (changes requested):
Some minor things i noticed.

arogge committed:
cmake: clarify stripping in bareos_add_compile_flags

arogge committed:
cmake: fix improper message type

arogge committed:
cmake: add bareos_compile_test()

extract repeating pattern in BareosCheckFunctions into a common
function.

arogge committed:
cmake: bump minimum required to 3.22

get rid of another conditional.

arogge committed:
cmake: add -static-libstdc++ for CXX linking only

arogge committed:
lib: always include bregex.h unconditionally

utilize the header checks that are present inside of bregex.h.

sebsura reviewed (approved):
Thanks for these awesome improvements!

sebsura committed:
apply check sources

BareosBot committed:
Update CHANGELOG.md
